File: src/main/java/org/organicdesign/testUtils/http/FakeHttpServletResponse.java
package org.organicdesign.testUtils.http;

import org.organicdesign.testUtils.http.FakeHttpServletRequest.Companion.Kv;
import javax.servlet.ServletOutputStream;
import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletResponse;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.stream.Collectors;

/**
 This mocks an HttpServletResponse.  It is a very raw and early version.
 */
public class FakeHttpServletResponse implements HttpServletResponse {
    private int status = 0;
    public String redirect = null;
    private Locale locale;
    private List<Map.Entry<String, String>> headers = new ArrayList<>();
    private String contentType = null;
    private boolean committed = false;
    private FakeServletOutputStream outputStream = new FakeServletOutputStream();

    @Override public Collection<String> getHeaderNames() {
        return headers.stream()
                      .map(Map.Entry::getKey)
                      .collect(Collectors.toSet());
    }
    @Override public String getHeader(String s) {
        Iterator<String> iter = getHeaders(s).iterator();
        return (iter.hasNext()) ? iter.next()
                                : null;
    }
    @Override public Collection<String> getHeaders(String s) {
        List<String> values = new ArrayList<>();
        if (s != null) {
            for (Map.Entry<String, String> head : headers) {
                if (s.equalsIgnoreCase(head.getKey())) {
                    values.add(head.getValue());
                }
            }
        }
        return values;
    }
    @Override public boolean containsHeader(String s) {
        if (s != null) {
            for (Map.Entry<String, String> head : headers) {
                if (s.equalsIgnoreCase(head.getKey())) {
                    return true;
                }
            }
        }
        return false;
    }
    @Override public void setHeader(String s, String s1)  {
        if (s != null) {
            for (int i = 0; i < headers.size(); i++) {
                Map.Entry<String, String> head = headers.get(i);
                if (s.equalsIgnoreCase(head.getKey())) {
                    headers.set(i, new Kv(s, s1));
                    return;
                }
            }
            addHeader(s, s1);
        }
    }
    @Override public void addHeader(String s, String s1)  {
        headers.add(new Kv(s, s1));
    }
    @Override public void setDateHeader(String s, long l)  {
        setHeader(s, String.valueOf(l));
    }
    @Override public void addDateHeader(String s, long l)  {
        addHeader(s, String.valueOf(l));
    }
    @Override public void setIntHeader(String s, int i)  {
        setHeader(s, String.valueOf(i));
    }
    @Override public void addIntHeader(String s, int i)  {
        addHeader(s, String.valueOf(i));
    }

    @Override public void addCookie(Cookie cookie) {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public String encodeURL(String s) {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public String encodeRedirectURL(String s) {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Deprecated
    @Override public String encodeUrl(String s) {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Deprecated
    @Override public String encodeRedirectUrl(String s) {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public void sendError(int i, String s) {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public void sendError(int i) {
        if (committed) {
            throw new IllegalStateException("Response already committed.");
        }
        status = i;
        contentType = "text/html";
        committed = true;
        // TODO: "clear the buffer"
    }
    @Override public void sendRedirect(String s) { redirect = s; }
    @Override public void setStatus(int i)  { status = i; }
    @Deprecated
    @Override public void setStatus(int i, String s)  { status = i; }
    @Override public int getStatus() { return status; }
    @Override public String getCharacterEncoding() {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public String getContentType() { return contentType; }
    @Override public ServletOutputStream getOutputStream() {
        return outputStream;
    }
    @Override public PrintWriter getWriter() {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public void setCharacterEncoding(String s) {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public void setContentLength(int i) {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public void setContentLengthLong(long l) {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public void setContentType(String s) { contentType = s; }
    @Override public void setBufferSize(int i) {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public int getBufferSize() {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public void flushBuffer() {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public void resetBuffer() {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public boolean isCommitted() { return committed; }
    @Override public void reset() {
        throw new UnsupportedOperationException("Not implemented");
    }
    @Override public void setLocale(Locale locale)  { this.locale = locale; }
    @Override public Locale getLocale() { return locale; }

    public static FakeHttpServletResponse httpServletResponse() {
        return new FakeHttpServletResponse() { };
    }
}

File: src/main/java/org/organicdesign/testUtils/http/FakeHttpServletResponse.kt
package org.organicdesign.testUtils.http

import jakarta.servlet.http.Cookie
import jakarta.servlet.http.HttpServletResponse
import org.organicdesign.indented.IndentedStringable
import org.organicdesign.indented.StringUtils.classFieldsK
import java.io.PrintWriter
import java.util.*

/**
 * This mocks an HttpServletResponse.  It is a very raw and early version.
 */
class FakeHttpServletResponse : HttpServletResponse, IndentedStringable {

    override fun indentedStr(indent: Int, singleLine: Boolean): String =
        classFieldsK(indent, "FakeHttpServletResponse",
                     listOf("status" to status,
                            "committed" to committed,
                            "redirect" to redirect,
                            "contentType" to contentType,
                            "encoding" to encoding,
                            "locale" to locale,
                            "cookies" to cookies.map{ CookiePrinter(it) },
                            "headers" to headers,
                            "outputStream" to outputStream
                     ).filter { it.second != null },
                     singleLine)

    override fun toString(): String = indentedStr(0)

    private var status = 0
    override fun setStatus(i: Int) { status = i }
    @Deprecated("")
    override fun setStatus(i: Int, s: String) { status = i }
    override fun getStatus(): Int = status

    var redirect: String? = null

    private var locale: Locale? = null
    override fun getLocale(): Locale? = locale
    override fun setLocale(locale: Locale) { this.locale = locale }

    private var contentType: String? = null
    override fun getContentType(): String? = contentType
    override fun setContentType(s: String) {
        contentType = s
        val charsetIdx = s.indexOf("charset=")
        if (charsetIdx > -1) {
            encoding = s.substringAfter("charset=").trim()
        }
    }

    private var committed = false
    override fun isCommitted(): Boolean = committed

    private var encoding: String? = null
    override fun getCharacterEncoding(): String? = encoding
    override fun setCharacterEncoding(s: String?) { encoding = s }

    val cookies: MutableList<Cookie> = mutableListOf()
    override fun addCookie(cookie: Cookie) { cookies.add(cookie) }

    private val outputStream = FakeServletOutputStream()
    override fun getOutputStream(): FakeServletOutputStream = outputStream
    override fun getWriter(): PrintWriter = PrintWriter(outputStream.stringWriter)
    private var bufferSize: Int = DEFAULT_BUFFER_SIZE
    override fun getBufferSize(): Int = bufferSize
    override fun setBufferSize(i: Int) { bufferSize = i }
    override fun flushBuffer() { }
    override fun resetBuffer() { }
    override fun reset() { }

    private val headers: MutableList<Map.Entry<String, String>> = mutableListOf()

    override fun getHeaderNames(): Collection<String> = headers.map{ it.key }.toSet()

    override fun getHeader(s: String): String? {
        val iter = getHeaders(s).iterator()
        return if (iter.hasNext())
            iter.next()
        else
            null
    }

    override fun getHeaders(s: String?): Collection<String> {
        val values: MutableList<String> = mutableListOf()
        if (s != null) {
            for (head in headers) {
                if (s.equals(head.key, ignoreCase = true)) {
                    values.add(head.value)
                }
            }
        }
        return values
    }

    override fun containsHeader(s: String?): Boolean {
        if (s != null) {
            for (head in headers) {
                if (s.equals(head.key, ignoreCase = true)) {
                    return true
                }
            }
        }
        return false
    }

    override fun setHeader(s: String?, s1: String) {
        if (s != null) {
            for (i in headers.indices) {
                val head = headers[i]
                if (s.equals(head.key, ignoreCase = true)) {
                    headers[i] = Kv(s, s1)
                    return
                }
            }
            addHeader(s, s1)
        }
    }

    override fun addHeader(s: String, s1: String) {
        headers.add(Kv(s, s1))
    }

    override fun setDateHeader(s: String, l: Long) {
        setHeader(s, l.toString())
    }

    override fun addDateHeader(s: String, l: Long) {
        addHeader(s, l.toString())
    }

    override fun setIntHeader(s: String, i: Int) {
        setHeader(s, i.toString())
    }

    override fun addIntHeader(s: String, i: Int) {
        addHeader(s, i.toString())
    }

    override fun encodeRedirectURL(s: String): String? {
        redirect = encodeURL(s)
        return redirect
    }
    @Deprecated(replaceWith = ReplaceWith("encodeRedirectURL(s)"),
                message = "replaced with encodeRedirectURL(String url)")
    override fun encodeRedirectUrl(s: String): String? = encodeRedirectURL(s)

    /**
     * This is only a correct implementation if we assume that the browser supports cookies, or session
     * tracking is turned off
     */
    override fun encodeURL(s: String): String = s

    @Deprecated(replaceWith = ReplaceWith("encodeURL(s)"),
                message = "encodeURL(s)")
    override fun encodeUrl(s: String): String = encodeURL(s)

    override fun sendError(i: Int, s: String) {
        if (committed) {
            throw IllegalStateException("Response already committed.")
        }
        status = i
        contentType = "text/html"
        committed = true
        // TODO: "clear the buffer"
        // TODO: Where does the string parameter go?
    }

    override fun sendError(i: Int) = sendError(i, "")

    override fun sendRedirect(s: String) { redirect = s }

    override fun setContentLength(i: Int) {
        throw UnsupportedOperationException("Not implemented")
    }

    override fun setContentLengthLong(l: Long) {
        throw UnsupportedOperationException("Not implemented")
    }

    companion object {
        @JvmStatic
        fun httpServletResponse(): FakeHttpServletResponse {
            return FakeHttpServletResponse()
        }
    }
}
